lib/seahorn/BvOpSem2.cc

```diff
@@ -835,13 +835,128 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
+      Expr is_underflow =
+          m_ctx.alu().IsUsubNoUnderflow(op0, op1, ty->getScalarSizeInBits());
+      Expr is_carry = m_ctx.alu().doNot(is_underflow, 1 /* bitwidth */);
+      Expr carry_1bv = m_ctx.alu().boolToBv1(is_carry);
+      setValue(I, createArithmeticWithOverflowRecord(sub_res, carry_1bv,
+                                                     getCarryBitPadWidth(I)));
+    } break;
+    case Intrinsic::smul_with_overflow: {
+      Type *ty = I.getOperand(0)->getType();
+      Expr op0 = lookup(*I.getOperand(0));
+      Expr op1 = lookup(*I.getOperand(1));
+
+      // assume doMul has same bitwidth as Op0
+      Expr mul_res = m_ctx.alu().doMul(op0, op1, ty->getScalarSizeInBits());
+      Expr is_overflow =
+          m_ctx.alu().IsSmulNoOverflow(op0, op1, ty->getScalarSizeInBits());
+      Expr is_underflow =
+          m_ctx.alu().IsBmulNoUnderflow(op0, op1, ty->getScalarSizeInBits());
+      Expr is_carry = m_ctx.alu().doNot(
+          m_ctx.alu().doAnd(is_overflow, is_underflow, 1 /* bitwidth */),
+          1 /* bitwidth*/);
+      Expr carry_1bv = m_ctx.alu().boolToBv1(is_carry);
+      setValue(I, createArithmeticWithOverflowRecord(mul_res, carry_1bv,
+                                                     getCarryBitPadWidth(I)));
+    } break;
+    case Intrinsic::umul_with_overflow: {
+      Type *ty = I.getOperand(0)->getType();
+      Expr op0 = lookup(*I.getOperand(0));
+      Expr op1 = lookup(*I.getOperand(1));
+
+      // assume doMul has same bitwidth as Op0
+      Expr mul_res = m_ctx.alu().doMul(op0, op1, ty->getScalarSizeInBits());
+      Expr is_overflow =
+          m_ctx.alu().IsUmulNoOverflow(op0, op1, ty->getScalarSizeInBits());
+      Expr is_carry = m_ctx.alu().doNot(
+          m_ctx.alu().doAnd(is_overflow, is_overflow, 1 /* bitwidth */),
+          1 /* bitwidth*/);
+      Expr carry_1bv = m_ctx.alu().boolToBv1(is_carry);
+      setValue(I, createArithmeticWithOverflowRecord(mul_res, carry_1bv,
+                                                     getCarryBitPadWidth(I)));
+    } break;
     default:
       // interpret by non-determinism (and a warning)
       if (!I.getType()->isVoidTy())
         setValue(I, Expr());
     }
   }
 
+  unsigned getCarryBitPadWidth(IntrinsicInst &I) {
+    const DataLayout &DL = m_sem.getDataLayout();
+    Type *curTy = I.getType();
+    auto *STy = dyn_cast<StructType>(curTy);
+    const StructLayout *SL = DL.getStructLayout(STy);
+    auto struc_size_in_bits = SL->getSizeInBits();
+    auto offset_in_bits = SL->getElementOffsetInBits(1);
+    auto carry_size_in_bits = (STy->getElementType(1))->getScalarSizeInBits();
+    return struc_size_in_bits - offset_in_bits - carry_size_in_bits;
+  }
+
+  Expr createArithmeticWithOverflowRecord(Expr &opResult, Expr &carryBit,
+                                          unsigned carryBitPadwidth) {
+    Expr carry = m_ctx.alu().Concat(m_ctx.alu().si(0U, carryBitPadwidth), carryBit);
```

Comment:
concat needs to know the bit-width of operands. Unfortunately, the width needs to be passed explicitly because it cannot be extracted from the operands themselves.